File: ocaml/cookbook/custom_tools.ml
(* Cookbook: Custom Tool Registration
   Demonstrates defining and registering tools that the Copilot agent
   can invoke during a session. *)

open Copilot_sdk

let () =
  Lwt_main.run
    (let open Lwt.Syntax in
     let client = Client.create () in
     let* () = Client.start client in

     (* Define a weather tool using the helpers *)
     let params =
       Tools.make_object_schema
         ~properties:
           [ Tools.make_string_param ~name:"location" ~description:"City name" ()
           ; Tools.make_string_param ~name:"units" ~description:"Temperature units (celsius/fahrenheit)" ()
           ]
         ~required:["location"]
         ()
     in
     let weather_def, weather_handler =
       Tools.define_tool
         ~name:"get_weather"
         ~description:"Get current weather for a location"
         ~parameters:params
         (fun invocation ->
           let open Yojson.Safe.Util in
           let location =
             try invocation.arguments |> member "location" |> to_string
             with _ -> "Unknown"
           in
           let result =
             Printf.sprintf
               "{\"location\": \"%s\", \"temperature\": 22, \"condition\": \"sunny\"}"
               location
           in
           Lwt.return (Types.make_tool_result result))
     in

     (* Define a calculator tool *)
     let calc_def, calc_handler =
       Tools.define_tool
         ~name:"calculate"
         ~description:"Evaluate a math expression"
         ~parameters:
           (Tools.make_object_schema
              ~properties:
                [ Tools.make_string_param
                    ~name:"expression"
                    ~description:"Math expression to evaluate"
                    ()
                ]
              ~required:["expression"]
              ())
         (fun invocation ->
           let open Yojson.Safe.Util in
           let expr =
             try invocation.arguments |> member "expression" |> to_string
             with _ -> "0"
           in
           (* Simple evaluation placeholder *)
           let result = Printf.sprintf "Result of '%s' = 42" expr in
           Lwt.return (Types.make_tool_result result))
     in

     (* Create session with both tools *)
     let config =
       { (Types.default_session_config ()) with
         tools = [ weather_def; calc_def ]
       }
     in
     let* session = Client.create_session ~config client in

     (* Register handlers *)
     Tools.register_tools session [ (weather_def, weather_handler); (calc_def, calc_handler) ];
     Session.register_permission_handler session Tools.approve_all;

     (* Use the tools *)
     let msg = Types.make_message "What is the weather in Paris? Also calculate 2+2." in
     let* _response = Session.send_and_wait session msg in

     let* () = Session.destroy session in
     let* () = Client.stop client in
     Lwt.return_unit)

File: julia/cookbook/custom_tools.jl
# Cookbook: Custom Tools
# Demonstrates registering tools that the assistant can call.

push!(LOAD_PATH, joinpath(@__DIR__, ".."))
using CopilotSDK

function custom_tools_example()
    # Define a calculator tool using the do-block syntax
    calc_tool = define_tool("calculate", "Evaluate a math expression";
        parameters=Dict{String, Any}(
            "type" => "object",
            "properties" => Dict{String, Any}(
                "expression" => Dict{String, Any}(
                    "type" => "string",
                    "description" => "A Julia math expression to evaluate",
                ),
            ),
            "required" => ["expression"],
        ),
    ) do inv
        expr = get(inv.arguments, "expression", "0")
        try
            result = eval(Meta.parse(expr))
            ToolResult(text_result_for_llm="Result: $result")
        catch e
            ToolResult(
                text_result_for_llm="Error: $(sprint(showerror, e))",
                result_type=TOOL_FAILURE,
            )
        end
    end

    # Define a greeting tool using positional syntax
    greet_tool = define_tool("greet", "Greet a user by name",
        inv -> ToolResult(text_result_for_llm="Hello, $(get(inv.arguments, "name", "World"))!"))

    client = CopilotClient(log_level=LOG_ERROR)
    start!(client)

    session = create_session(client;
        model="gpt-4",
        tools=[calc_tool, greet_tool],
        on_permission_request=approve_all,
    )

    done = Channel{Bool}(1)
    on(session) do event
        if event.type == "assistant.message"
            println("Assistant: ", get(event.data, "content", ""))
        elseif event.type == "session.idle"
            try put!(done, true) catch; end
        end
    end

    send(session, "Calculate 42 * 17 + 3, then greet Julia.")

    CopilotSDK.timedwait(60.0) do
        isready(done)
    end

    disconnect(session)
    stop!(client)
end

custom_tools_example()
